server/lib/two-factor-authentication/lib.ts

```diff
@@ -0,0 +1,168 @@
+import { ApolloError } from 'apollo-server-errors';
+import { Request } from 'express';
+import { isNil } from 'lodash';
+
+import User from '../../models/User';
+import cache from '../cache';
+
+import totp from './totp';
+
+const DEFAULT_TWO_FACTOR_AUTH_SESSION_DURATION = 600; // 10min
+
+type ValidateRequestOptions = {
+  // require user configured 2FA
+  requireTwoFactorAuthEnabled?: boolean;
+  // always ask for a token when using 2FA
+  alwaysAskForToken?: boolean;
+  // duration which we wont require a token after a successful use
+  sessionDuration?: number;
+  // identifier for the session
+  sessionKey?: (() => string) | string;
+};
+
+export enum TwoFactorMethod {
+  TOTP = 'totp',
+}
+
+export const TwoFactorAuthenticationHeader = 'x-two-factor-authentication';
+
+export const SupportedTwoFactorMethods = [TwoFactorMethod.TOTP];
+
+export type Token = {
+  type: TwoFactorMethod;
+  code: string;
+};
+
+export interface TwoFactorAuthProvider {
+  validateToken(user: typeof User, token: Token): Promise<void>;
+}
+
+export const providers: { [method in TwoFactorMethod]: TwoFactorAuthProvider } = {
+  [TwoFactorMethod.TOTP]: totp,
+};
+
+function getTwoFactorAuthTokenFromRequest(req: Request): Token {
+  const header = req.get(TwoFactorAuthenticationHeader);
+  if (!header) {
+    return null;
+  }
+
+  const parts = header.split(' ');
+  const type = parts[0] as TwoFactorMethod;
+  const code = parts[1];
+
+  return {
+    type,
+    code,
+  };
+}
+
+async function validateToken(user: typeof User, token: Token): Promise<void> {
+  if (!SupportedTwoFactorMethods.includes(token.type)) {
+    throw new Error(`Unsupported 2FA type ${token.type}`);
+  }
+
+  return providers[token.type].validateToken(user, token);
+}
+
+const DefaultValidateRequestOptions: ValidateRequestOptions = {
+  requireTwoFactorAuthEnabled: false,
+  alwaysAskForToken: false,
+  sessionDuration: DEFAULT_TWO_FACTOR_AUTH_SESSION_DURATION,
+};
+
+function getSessionKey(req: Request, options: ValidateRequestOptions) {
+  if (typeof options.sessionKey === 'function') {
+    return options.sessionKey();
+  } else if (typeof options.sessionKey === 'string') {
+    return options.sessionKey;
+  } else {
+    const userId = req?.remoteUser?.id;
+    const sessionId = req.jwtPayload?.sessionId;
+    return `${userId}_2fa_${sessionId}`;
+  }
```

Comment:
:ice_cream: I suggest adding a prefix to all cache keys, mostly to facilitate the exploration in Redis GUI tools but that could also be used in the future to easily invalidate all 2FA for a given user.

```suggestion
  const userId = req.remoteUser.id;
  if (typeof options.sessionKey === 'function') {
    return `2fa:${userId}:${options.sessionKey()}`;
  } else if (typeof options.sessionKey === 'string') {
    return `2fa:${userId}:${options.sessionKey}`;
  } else {
    const sessionId = req.jwtPayload.sessionId;
    return `2fa:${userId}:default_${sessionId}`;
  }
```

Reply:
Thanks for catching this, enforcing the userID in the key prevents someone passing a key that is valid to all users.